src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+            }
+        }
+
+        spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_type(type));
+        return id;
+    }
+
+    void ShaderProgram::attach_shader(GLuint const shader) const {
+        glAttachShader(m_program_name, shader);
+        glDeleteShader(shader);
+    }
+
+    void ShaderProgram::link_program() const {
+        glLinkProgram(m_program_name);
+
+        GLint success;
+        glGetProgramiv(m_program_name, GL_LINK_STATUS, &success);
+        if (not success) {
+            auto message = std::string{};
+            GLint len;
+            glGetProgramiv(m_program_name, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow<usize>(len - 2));
+            glGetProgramInfoLog(m_program_name, len - 1, nullptr, message.data());
+            spdlog::critical("ERROR::PROGRAMM::LINK_FAILED -> {}\n", message.data());
+            glDeleteProgram(m_program_name);
+            throw GLError(GLErrorType::FailedToLinkShaderProgram, message);
+        }
+    }
+
+    ShaderProgram::ShaderProgram(
+            std::string_view const vertex_source,
+            std::string_view const geometry_source,
+            std::string_view const fragment_source
+    )
+        : m_program_name{ 0 } {
+        m_program_name = { glCreateProgram() };
+
+
+        auto const vertex_shader = compile(Type::Vertex, vertex_source);
+        attach_shader(vertex_shader);
+
+        if (not geometry_source.empty()) {
+            auto const geometry_shader = compile(Type::Geometry, geometry_source);
+            attach_shader(geometry_shader);
+        }
+
+        auto const fragment_shader = compile(Type::Fragment, fragment_source);
+        attach_shader(fragment_shader);
+
+        link_program();
+    }
+
+    ShaderProgram::ShaderProgram(std::string_view const vertex_source, std::string_view const fragment_source)
+        : ShaderProgram(vertex_source, "", fragment_source) { }
+
+    ShaderProgram::ShaderProgram(ShaderProgram&& other) noexcept
+        : m_program_name{ std::exchange(other.m_program_name, 0) } { }
+
+    ShaderProgram& ShaderProgram::operator=(ShaderProgram&& other) noexcept {
+        std::swap(m_program_name, other.m_program_name);
+        return *this;
+    }
+
+    ShaderProgram::~ShaderProgram() {
+        glDeleteProgram(m_program_name);
+    }
+
+
+    void ShaderProgram::use() {
+        glUseProgram(m_program_name);
+    }
```

Comment:
Do we also want an `unuse()` function? Or maybe call them `bind()` and `unbind()` to be consistent to the buffer objects.